File: src/main/java/de/hdawg/rankinginfo/service/RankingService.java
package de.hdawg.rankinginfo.service;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.Month;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.TreeMap;
import java.util.stream.Collectors;

import org.jspecify.annotations.Nullable;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import de.hdawg.rankinginfo.domain.Ranking;
import de.hdawg.rankinginfo.domain.RankingCoding;
import de.hdawg.rankinginfo.repository.ImportHistoryRepository;
import de.hdawg.rankinginfo.repository.RankingQueryFilter;
import de.hdawg.rankinginfo.repository.RankingRepository;

/// Business-logic facade over [RankingRepository] for browsing rankings: translates the
/// web/API-facing [RankingFilter] into a [RankingQueryFilter] the repository understands, and
/// exposes supporting lookups (available quarters, federations, previous-quarter positions, last
/// import timestamp).
///
/// Age-group slugs (e.g. `m00`, `mu18`) are translated to/from stored `age_group` values by
/// [#toAgeGroupSlug(String, String)] and [#slugToAgeGroup(String)].
@Service
@Transactional(readOnly = true)
public class RankingService {

  private static final String AGE_GROUP_M00 = "m00";
  private static final String AGE_GROUP_W00 = "w00";
  private static final String AGE_GROUP_OVERALL = "overall";

  private record YobAgeFlags(boolean yobRanking, boolean ageGroupRanking) {}

  private static final Map<String, YobAgeFlags> AGE_GROUP_OPTION_FILTERS =
      Map.of(
          "only_yob", new YobAgeFlags(true, false),
          "include_younger", new YobAgeFlags(false, false));

  private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("dd.MM.yyyy");

  private final RankingRepository rankingRepository;
  private final ImportHistoryRepository importHistoryRepository;

  public RankingService(
      RankingRepository rankingRepository, ImportHistoryRepository importHistoryRepository) {
    this.rankingRepository = rankingRepository;
    this.importHistoryRepository = importHistoryRepository;
  }

  /// Lists every quarter that has ranking data, grouped by year.
  ///
  /// Each stored ranking date is the first day of the following quarter (e.g. data for Q1 is
  /// dated April 1st); entries are adjusted back by one day before being grouped and formatted,
  /// except the synthetic December 31st year-end ranking date, which is skipped entirely.
  ///
  /// @return years (as strings, descending) mapped to their available quarters
  @Cacheable("available_quarters")
  public Map<String, List<QuarterEntry>> fetchAvailableQuarters() {
    var years = new TreeMap<String, List<QuarterEntry>>(Comparator.reverseOrder());
    for (var date : rankingRepository.findDistinctDatesDesc()) {
      if (date.getMonthValue() == Month.DECEMBER.getValue() && date.getDayOfMonth() == 31) continue;
      var adjusted = date.minusDays(1);
      years
          .computeIfAbsent(String.valueOf(adjusted.getYear()), k -> new ArrayList<>())
          .add(new QuarterEntry(adjusted.format(DATE_FORMATTER), date.toString()));
    }
    return years;
  }

  /// Lists every distinct federation abbreviation present in the stored rankings.
  ///
  /// @return distinct federation abbreviations
  @Cacheable("federations")
  public List<String> fetchFederations() {
    return rankingRepository.findDistinctFederations();
  }

  /// Fetches one page of rankings matching `filter`.
  ///
  /// @param filter the web/API-facing filter criteria
  /// @param page zero-based page index
  /// @param perPage page size
  /// @return the matching page of rankings
  public Page<Ranking> findFilteredRankings(RankingFilter filter, int page, int perPage) {
    return rankingRepository.findFiltered(buildQueryFilter(filter), page, perPage);
  }

  /// Looks up each given player's ranking position in the quarter immediately preceding
  /// `filter`'s quarter, for computing position-change indicators.
  ///
  /// @param filter the filter whose quarter to look up the *previous* quarter for; its other
  ///     criteria (age group, federation, club) are reused as-is
  /// @param dtbIds the players to look up; returns an empty map if this is empty
  /// @return each looked-up player's previous-quarter ranking position, keyed by DTB ID; empty
  ///     if `filter`'s quarter has no preceding quarter with data
  public Map<Integer, Integer> findPreviousPositions(RankingFilter filter, List<Integer> dtbIds) {
    if (dtbIds.isEmpty()) return Map.of();
    var quarterDate = LocalDate.parse(filter.quarter());
    var prevDate =
        rankingRepository.findDistinctDatesDesc().stream()
            .filter(d -> d.isBefore(quarterDate))
            .findFirst()
            .orElse(null);
    if (prevDate == null) return Map.of();
    var prevFilter =
        buildQueryFilter(filter.withQuarter(prevDate.toString())).withDtbIds(dtbIds);
    return rankingRepository.findFiltered(prevFilter).stream()
        .collect(Collectors.toMap(Ranking::dtbId, Ranking::rankingPosition));
  }

  /// Returns the timestamp of the most recent successful import, if any.
  ///
  /// @return the latest import timestamp, or `null` if nothing has been imported yet
  @Cacheable("max_imported_at")
  @Nullable
  public LocalDateTime maxImportedAt() {
    return importHistoryRepository.findMaxImportedAt();
  }

  private static RankingQueryFilter buildQueryFilter(RankingFilter filter) {
    var quarter = LocalDate.parse(filter.quarter());
    var slug = filter.ageGroupSlug();
    var ageGroup = slugToAgeGroup(slug);
    var isMale = slug.startsWith("m");
    var isAdult = AGE_GROUP_M00.equals(ageGroup) || AGE_GROUP_W00.equals(ageGroup);

    int dtbIdStart = isMale ? RankingCoding.MALE_DTB_ID_START : RankingCoding.FEMALE_DTB_ID_START;
    int dtbIdEnd = isMale ? RankingCoding.MALE_DTB_ID_END : RankingCoding.FEMALE_DTB_ID_END;

    boolean yobRanking;
    boolean ageGroupRanking;
    if (isAdult) {
      yobRanking = false;
      ageGroupRanking = false;
    } else if (filter.ageGroupOptions() == null) {
      int n = parseAgeGroupNumber(ageGroup);
      yobRanking = (n % 2 != 0);
      ageGroupRanking = (n % 2 == 0);
    } else {
      var flags =
          AGE_GROUP_OPTION_FILTERS.getOrDefault(
              filter.ageGroupOptions(), new YobAgeFlags(false, false));
      yobRanking = flags.yobRanking();
      ageGroupRanking = flags.ageGroupRanking();
    }

    var fedValue = filter.federation();
    var fed = fedValue != null && !fedValue.isBlank() ? fedValue : null;
    var clubValue = filter.club();
    var club = clubValue != null && !clubValue.isBlank() ? clubValue : null;

    return new RankingQueryFilter(
        quarter,
        ageGroup,
        dtbIdStart,
        dtbIdEnd,
        yobRanking,
        ageGroupRanking,
        filter.yearEnd() && quarter.getMonthValue() == Month.JANUARY.getValue(),
        fed,
        club,
        null);
  }

  /// Builds the URL/API age-group slug (e.g. `m00`, `mu18`) for a gender category and, for
  /// junior categories, an optional age-group label.
  ///
  /// @param gender one of `Herren`, `Damen`, `Junioren`, `Juniorinnen`; anything else yields
  ///     `overall`
  /// @param ageGroup for `Junioren`/`Juniorinnen`, the age-group label (e.g. `U18`); `null` or
  ///     blank yields the `overall` slug for that gender
  /// @return the age-group slug
  public static String toAgeGroupSlug(String gender, @Nullable String ageGroup) {
    return switch (gender) {
      case "Herren" -> "m00";
      case "Damen" -> "w00";
      case "Junioren" ->
          (ageGroup == null || ageGroup.isBlank())
              ? AGE_GROUP_OVERALL
              : "m" + ageGroup.toLowerCase(Locale.ROOT);
      case "Juniorinnen" ->
          (ageGroup == null || ageGroup.isBlank())
              ? AGE_GROUP_OVERALL
              : "w" + ageGroup.toLowerCase(Locale.ROOT);
      default -> AGE_GROUP_OVERALL;
    };
  }

  private static String slugToAgeGroup(String slug) {
    if (AGE_GROUP_M00.equals(slug)) return AGE_GROUP_M00;
    if (AGE_GROUP_W00.equals(slug)) return AGE_GROUP_W00;
    return "U" + slug.substring(2).toUpperCase(Locale.ROOT);
  }

  private static int parseAgeGroupNumber(String ageGroup) {
    if (!ageGroup.startsWith("U")) return 0;
    try {
      return Integer.parseInt(ageGroup.substring(1));
    } catch (NumberFormatException _) {
      return 0;
    }
  }
}

File: src/main/kotlin/de/hdawg/rankinginfo/service/RankingService.kt
package de.hdawg.rankinginfo.service

import de.hdawg.rankinginfo.domain.Ranking
import de.hdawg.rankinginfo.repository.ImportHistoryRepository
import de.hdawg.rankinginfo.repository.RankingQueryFilter
import de.hdawg.rankinginfo.repository.RankingRepository
import org.springframework.cache.annotation.Cacheable
import org.springframework.data.domain.Page
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * Parameters for filtering a ranking list query.
 *
 * @property quarter ISO date string for the ranking period (e.g. "2024-01-01")
 * @property ageGroupSlug slug encoding gender and age group (e.g. "m12", "wu18", "m00")
 * @property ageGroupOptions optional variant selector: "only_yob" or "include_younger"
 * @property federation optional federation abbreviation filter
 * @property club optional club name filter (case-insensitive substring)
 * @property yearEnd when true, selects year-end rankings (relevant for January quarters only)
 */
data class RankingFilter(
    val quarter: String,
    val ageGroupSlug: String,
    val ageGroupOptions: String? = null,
    val federation: String? = null,
    val club: String? = null,
    val yearEnd: Boolean = false,
)

/**
 * Service for querying and filtering ranking data.
 *
 * All read operations are transactional (read-only). Frequently accessed lookups
 * (available quarters, federations) are cached via Spring's caching abstraction.
 */
@Service
@Transactional(readOnly = true)
class RankingService(
    private val rankingRepository: RankingRepository,
    private val importHistoryRepository: ImportHistoryRepository,
) {
    companion object {
        private val AGE_GROUP_OPTION_FILTERS =
            mapOf(
                "only_yob" to Pair(true, false),
                "include_younger" to Pair(false, false),
            )
        private val DATE_FORMATTER = DateTimeFormatter.ofPattern("dd.MM.yyyy")

        private fun slugToAgeGroup(slug: String): String =
            when {
                slug == "m00" -> "m00"
                slug == "w00" -> "w00"
                else -> "U${slug.substring(2).uppercase()}"
            }

        private fun buildQueryFilter(filter: RankingFilter): RankingQueryFilter {
            val quarter = LocalDate.parse(filter.quarter)
            val slug = filter.ageGroupSlug
            val ageGroup = slugToAgeGroup(slug)
            val isMale = slug.startsWith("m")
            val isAdult = ageGroup == "m00" || ageGroup == "w00"

            val (dtbIdStart, dtbIdEnd) =
                if (isMale) Pair(10_000_000, 19_999_999) else Pair(20_000_000, 29_999_999)

            val (yobRanking, ageGroupRanking) =
                if (isAdult) {
                    Pair(false, false)
                } else {
                    when (val opts = filter.ageGroupOptions) {
                        null -> {
                            val n = ageGroup.removePrefix("U").toIntOrNull() ?: 0
                            if (n % 2 == 0) Pair(false, true) else Pair(true, false)
                        }
                        else -> AGE_GROUP_OPTION_FILTERS[opts] ?: Pair(false, false)
                    }
                }

            val isJanuary = quarter.monthValue == 1
            return RankingQueryFilter(
                date = quarter,
                ageGroup = ageGroup,
                dtbIdStart = dtbIdStart,
                dtbIdEnd = dtbIdEnd,
                yobRanking = yobRanking,
                ageGroupRanking = ageGroupRanking,
                yearEndRanking = filter.yearEnd && isJanuary,
                federation = filter.federation?.takeIf { it.isNotBlank() },
                club = filter.club?.takeIf { it.isNotBlank() },
            )
        }
    }

    /**
     * Returns all available ranking quarters grouped by year, sorted descending.
     *
     * Dates with month=12/day=31 (year-end marker rows) are skipped.
     * The result is cached under `available_quarters`.
     */
    @Cacheable("available_quarters")
    fun fetchAvailableQuarters(): Map<String, List<Pair<String, String>>> {
        val years = mutableMapOf<String, MutableList<Pair<String, String>>>()
        for (date in rankingRepository.findDistinctDatesDesc()) {
            if (date.monthValue == 12 && date.dayOfMonth == 31) continue
            val adjusted = date.minusDays(1)
            years
                .getOrPut(adjusted.year.toString()) { mutableListOf() }
                .add(adjusted.format(DATE_FORMATTER) to date.toString())
        }
        return years.entries.sortedByDescending { it.key.toInt() }.associate { it.key to it.value }
    }

    /** Returns all distinct federation names, sorted alphabetically. Cached under `federations`. */
    @Cacheable("federations")
    fun fetchFederations(): List<String> = rankingRepository.findDistinctFederations()

    /**
     * Returns a paginated page of rankings matching the given [filter].
     *
     * @param filter the ranking filter criteria
     * @param page 1-based page number
     * @param perPage number of results per page
     */
    fun findFilteredRankings(
        filter: RankingFilter,
        page: Int,
        perPage: Int,
    ): Page<Ranking> = rankingRepository.findFiltered(buildQueryFilter(filter), page, perPage)

    /**
     * Looks up the ranking positions from the quarter immediately preceding [filter]'s quarter
     * for the given set of DTB IDs.
     *
     * Returns an empty map when no previous quarter exists or [dtbIds] is empty.
     *
     * @param filter used to determine the current quarter and apply the same filter criteria
     * @param dtbIds DTB IDs to look up; typically the IDs on the current page
     * @return map of DTB ID to previous ranking position
     */
    fun findPreviousPositions(
        filter: RankingFilter,
        dtbIds: List<Int>,
    ): Map<Int, Int> {
        if (dtbIds.isEmpty()) return emptyMap()
        val prevDate =
            rankingRepository
                .findDistinctDatesDesc()
                .firstOrNull { it < LocalDate.parse(filter.quarter) } ?: return emptyMap()

        val prevFilter = buildQueryFilter(filter.copy(quarter = prevDate.toString())).copy(dtbIds = dtbIds)
        return rankingRepository.findFiltered(prevFilter).associate { it.dtbId to it.rankingPosition }
    }

    /** Returns the timestamp of the most recent completed import, or null if no imports exist. */
    fun maxImportedAt(): LocalDateTime? = importHistoryRepository.findMaxImportedAt()
}
